packages/shorebird_cli/lib/src/executables/devicectl/devicectl.dart

```diff
@@ -0,0 +1,311 @@
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:collection/collection.dart';
+import 'package:io/io.dart';
+import 'package:json_path/json_path.dart';
+import 'package:path/path.dart' as p;
+import 'package:scoped/scoped.dart';
+import 'package:shorebird_cli/src/executables/devicectl/apple_device.dart';
+import 'package:shorebird_cli/src/executables/devicectl/nserror.dart';
+import 'package:shorebird_cli/src/logger.dart';
+import 'package:shorebird_cli/src/process.dart';
+import 'package:shorebird_cli/src/third_party/flutter_tools/lib/flutter_tools.dart';
+import 'package:shorebird_code_push_client/shorebird_code_push_client.dart';
+
+typedef BundleId = String;
+
+/// {@template devicectl_exception}
+/// Thrown when a [Devicectl] command fails.
+/// {@endtemplate}
+class DevicectlException implements Exception {
+  /// {@macro devicectl_exception}
+  DevicectlException({
+    required this.message,
+    this.underlyingException,
+  });
+
+  /// A message describing this exception.
+  final String message;
+
+  /// The exception that caused this exception to be thrown, if any.
+  final Exception? underlyingException;
+
+  @override
+  String toString() => '''
+DevicectlException: $message
+Underlying exception: ${underlyingException ?? '(none)'}
+''';
+}
+
+/// A reference to a [Devicectl] instance.
+final devicectlRef = create(Devicectl.new);
+
+/// The [Devicectl] instance available in the current zone.
+Devicectl get devicectl => read(devicectlRef);
+
+/// A wrapper around the `devicectl` command.
+class Devicectl {
+  static const executableName = 'xcrun';
+  static const baseArgs = [
+    'devicectl',
+  ];
+
+  /// Whether the `devicectl` command is available.
+  Future<bool> _isAvailable() async {
+    try {
+      final result = await process.run(executableName, [
+        ...baseArgs,
+        '--version',
+      ]);
+      return result.exitCode == ExitCode.success.code;
+    } catch (_) {
+      return false;
+    }
+  }
+
+  /// Returns the first available iOS device, or the device with the given
+  /// [deviceId] if provided. Devices that are running iOS <17 are not
+  /// "CoreDevice"s and are not visible to devicectl.
+  Future<AppleDevice?> _deviceForLaunch({String? deviceId}) async {
+    final devices = await listAvailableIosDevices();
+
+    if (deviceId != null) {
+      return devices.firstWhereOrNull((d) => d.identifier == deviceId);
+    } else {
+      return devices.firstOrNull;
+    }
+  }
+
+  /// Whether we should use `devicectl` to install and launch the app on the
+  /// device with the given [deviceId], or the first available device we find if
+  /// [deviceId] is not provided.
+  Future<bool> shouldUseDevicectl({String? deviceId}) async {
```

Comment:
```suggestion
  Future<bool> isSupported({String? deviceId}) async {
```

nit but I feel this would read a bit better

Reply:
or maybe `isCoreDevice`?

Reply:
Yeah, I went through a few iterations of naming this and didn't land on anything I liked. Will do this.

Reply:
Done